In [55]:
import numpy as np
import matplotlib.pyplot as plt
import json

Compute the conditional entropy of the patterns
$$\begin{align}
    H(x_{t}|x_{t-1})&=-\sum_{(x_{t},x_{t-1})}p(x_{t},x_{t-1})\log_2\left(p(x_{t}|x_{t-1})\right)\nonumber\\
    &=-\sum_{(x_{t},x_{t-1})}p(x_{t},x_{t-1})\log_2\left(\frac{p(x_{t},x_{t-1})}{p(x_{t-1})}\right)\nonumber\\
    &=H(x_{t},x_{t-1})-H(x_{t-1}),
\end{align}$$
where $x_t$ is the state of the agent at time $t$.



In [56]:
# No. of agents
N = 11
# No. of available spots
s = 5

In [57]:
with open(f'graph_data_N{N:d}s{s:d}.json', 'r') as f:
  struct = json.load(f)

In [58]:
struct

{'0': {'pattern': ['0', '1', '1', '0', '1', '1', '0', '0', '1', '0', '0'],
  'neigh': ['0', '2', '4', '8'],
  'strat': {'0001': '1',
   '1110': '1',
   '1001': '0',
   '0011': '1',
   '1000': '1',
   '0110': '0',
   '1100': '0'}},
 '1': {'pattern': ['0', '1', '0', '0', '1', '1', '1', '0', '1', '0', '0'],
  'neigh': ['0', '1', '2'],
  'strat': {'000': '1',
   '111': '0',
   '100': '0',
   '110': '1',
   '011': '0',
   '001': '0'}},
 '2': {'pattern': ['0', '1', '0', '0', '0', '0', '1', '0', '1', '1', '1'],
  'neigh': ['0', '1', '2', '4', '9'],
  'strat': {'00000': '1',
   '11111': '0',
   '10000': '0',
   '00010': '0',
   '11001': '0',
   '11000': '1',
   '01111': '0',
   '11101': '1',
   '00111': '1',
   '00110': '0'}},
 '3': {'pattern': ['1', '0', '0', '1', '0', '0', '0', '1', '0', '1', '1'],
  'neigh': ['0', '1', '2', '4'],
  'strat': {'0000': '0',
   '1111': '0',
   '1000': '1',
   '0001': '0',
   '1100': '0',
   '0111': '1',
   '1110': '1',
   '0011': '1'}},
 '4': {'pattern': ['0', 

In [59]:
def distribution(pattern):
    freqs = {}
    for i in range(len(pattern)):
        x_t = pattern[i]
        freqs[x_t] = freqs.get(x_t, 0) + 1
    
    joint_freqs = {}
    for i in range(len(pattern) - 1):
        x_t = pattern[i + 1]
        x_t_1 = pattern[i]
        joint_freqs[x_t+x_t_1] = joint_freqs.get(x_t+x_t_1, 0) + 1
    x_t = pattern[0]
    x_t_1 = pattern[len(pattern)-1]
    joint_freqs[x_t+x_t_1] = joint_freqs.get(x_t+x_t_1, 0) + 1

    for k in freqs:
        freqs[k] /= len(pattern)
    for k in joint_freqs:
        joint_freqs[k] /= len(pattern)

    return freqs, joint_freqs

def cond_entropy(pattern):
    dist, joint_dist = distribution(pattern)
    H_X = -sum(p * np.log2(p) for p in dist.values() if p > 0)
    H_X_Y = -sum(p * np.log2(p) for p in joint_dist.values() if p > 0)
    return H_X_Y - H_X

In [60]:
for st in struct:
    cond_ent = cond_entropy(struct[st]['pattern'])
    n_neigh = len(struct[st]['neigh'])

    print(f'agent: {st}, cond_ent: {cond_ent:0.3f}, n_neigh: {n_neigh}')

agent: 0, cond_ent: 0.987, n_neigh: 4
agent: 1, cond_ent: 0.987, n_neigh: 3
agent: 2, cond_ent: 0.987, n_neigh: 5
agent: 3, cond_ent: 0.987, n_neigh: 4
agent: 4, cond_ent: 0.829, n_neigh: 4
agent: 5, cond_ent: 0.987, n_neigh: 4
agent: 6, cond_ent: 0.829, n_neigh: 4
agent: 7, cond_ent: 0.987, n_neigh: 4
agent: 8, cond_ent: 0.829, n_neigh: 5
agent: 9, cond_ent: 0.829, n_neigh: 5
agent: 10, cond_ent: 0.987, n_neigh: 4


agent: 0, cond_ent: 0.987, n_neigh: 4

agent: 1, cond_ent: 0.829, n_neigh: 3

agent: 2, cond_ent: 0.355, n_neigh: 3

agent: 3, cond_ent: 0.987, n_neigh: 4

agent: 4, cond_ent: 0.355, n_neigh: 1

agent: 5, cond_ent: 0.355, n_neigh: 1

agent: 6, cond_ent: 0.829, n_neigh: 3

agent: 7, cond_ent: 0.829, n_neigh: 3

agent: 8, cond_ent: 0.829, n_neigh: 3

agent: 9, cond_ent: 0.355, n_neigh: 3

agent: 10, cond_ent: 0.829, n_neigh: 3

In [61]:
0.986*10

9.86